token-client/src/main/java/com/sap/cloud/security/xsuaa/tokenflows/JwtBearerTokenFlow.java

```diff
@@ -41,17 +47,44 @@ public JwtBearerTokenFlow(@Nonnull OAuth2TokenService tokenService,
     }
 
     /**
-     * Sets the bearer token for the next execution.
+     * Sets the bearer token that should be exchanged for another JWT token.
      *
      * @param bearerToken
      *            - the bearer token.
      * @return this builder.
      */
     public JwtBearerTokenFlow token(String bearerToken) {
+        assertNotNull(bearerToken, "Bearer token must not be null.");
         this.bearerToken = bearerToken;
         return this;
     }
 
+    /**
+     * Sets the JWT token that should be exchanged for another JWT token.
+     * Use this setter if you want zid claim  from the token to be explicitly set in the X-zid header.
+     *
+     * @param token
+     *            - the Token.
+     * @return this builder.
+     */
+    public JwtBearerTokenFlow token(Token token) {
+        this.bearerToken = token.getTokenValue();
+        this.xZid = token.getZoneId();
+        return this;
+    }
+
+    /**
+     * Sets the zoneid of the tenant<br>
+     *
+     * @param zoneId
+     *            - the zoneId.
+     * @return this builder.
+     */
+    public JwtBearerTokenFlow zoneId(String zoneId) {
```

Comment:
If I give a zone ID here that is `!=` the ID contained within the `Token token` passed above, what would happen? Should we throw an exception in this case?

Reply:
if the zone id doesn't exists xsuaa would respond with 404 This `zoneId(String zoneId)` setter was meant to be used together with `token(String bearerToken)` as with this setter there's no option to set the `X-zid` header unless you use `token(Token token)` setter. I wouldn't restrict setting the zone id to the same value as in the token as I cannot enforce the same logic in case `token(String bearerToken)`  is used, moreover zone id header is still optional.